KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,652 @@
+
+The smart contract is mainly for recording the details, and the Core will execute the fund re-distributions.
+
+### Design decision
+
+#### KLAY transfer is not allowed via smart contracts
+
+As the balance of treasury funds keeps increasing for every block with the block reward its hard to keep track of the balances and rebalance token allocation. So smart contract will only record the rebalanced allocation and the core will execute the allocation reading from the contract.
+
+#### Approval of retiredAddress
+
+To record the addresses in a verifiable manner the addresses are verified in the contract by calling approve method. The retiredAddress can be a Contract address or a Externally Owned Account. If the retired address is a
+
+- EOA : EOA address can be verified when the account holder directly calls the approve function. `msg.sender == retiredAddress`
+- Contract : Contract address can be verified when the admin of the contract calls approve function. The smart contract calls the `getState()` function implemented in the retiredAddress contract to get the admin details. `msg.sender == admin`
+  - `getState()` funtion is implemented in Klaytn treasury contracts. It returns the adminList and quorom (min required admins to approve).
+  - Condition: Min required admins should approve the senderAddress contract.
+
+#### No Withdrawal
+
+Smart contract is not allowed to receive KLAY due to security reasons. So any funds send the contract will be reverted and withdraw function is not implemented.
+
+#### Finalize Contract
+
+Once the re-distribution a.k.a rebalance is executed by the Core, the status of the smart contract will be finalized and any modifications to the storage data will be restricted.
+
+## Backwards Compatibility
+
+<!-- All KIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The KIP must explain how the author proposes to deal with these incompatibilities. KIP submissions without a sufficient backwards compatibility treatise may be rejected outright. The authors should answer the question: "Does this KIP require a hard fork?" -->
+
+- The foundation should deploy new TreasuryRebalance contract to record the token redistribution.
+- To rebalance the funds and redistribute in a consistent manner the foundation should burn the designated funds before re-distribution.
+- This does not affect the backward compatibility as this a newly dpeloyed contract
+
+## Implementation
+
+<!--The implementations must be completed before any KIP is given status "Final", but it need not be completed before the KIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
+
+#### Example implementation
```

Comment:
@toniya-klaytn It would be good to have an interface of this specification rather than showing whole implementation.

Reply:
Shall I remove the implementation and add a interface ?

Reply:
Added the interface. The implementation link can be added once the code is merged in Klaytn [repo](https://github.com/klaytn/klaytn/pull/1803).